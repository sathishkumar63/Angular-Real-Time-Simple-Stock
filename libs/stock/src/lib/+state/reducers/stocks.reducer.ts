import { StocksAction, StocksActionTypes } from '../actions';

export interface StocksState {
  data: any[];
  loading: boolean;
  loaded: boolean;
}

export const initialState: StocksState = {
  data: null,
  loading: false,
  loaded: false,
};

export function stocksReducer(
  state: StocksState = initialState,
  action: StocksAction
): StocksState {
  switch (action.type) {
    case StocksActionTypes.GetStocks:
      return {
        ...state,
        loading: true,
        loaded: false,
      };
    case StocksActionTypes.GetStocksSuccess:
      return {
        ...state,
        data: action.payload,
        loading: false,
        loaded: true,
      };
    case StocksActionTypes.GetStocksFail:
      return {
        ...state,
        loading: false,
        loaded: false,
      };
    case StocksActionTypes.SetStocks:
      return {
        ...state,
        data: state.data?.map((item) =>
          item.symbol === action.payload.symbol
            ? { ...item, ...action.payload }
            : item
        ),
      };
    default:
      return state;
  }
}
