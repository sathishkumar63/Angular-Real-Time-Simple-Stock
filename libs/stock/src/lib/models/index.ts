export * from './stock.constant';
export * from './stock.interface';
export * from './stock.mock.data';
