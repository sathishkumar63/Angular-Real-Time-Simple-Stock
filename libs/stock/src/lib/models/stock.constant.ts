export const stocksSymbol: string[] = [
  'GOOG',
  'GOOGL',
  'MSFT',
  'AAPL',
  'TSLA',
  'AMZN',
  'UBER',
  'DIDI',
  'META',
  'NFLX',
  'CHIK',
  'GLD',
  'IAU',
  'BTC-USD',
  'ETH-USD',
  'DOGE-USD',
  'SHIB-USD',
  'ETH-CAD',
  'USDT-USD',
  'USDC-USD',
];
