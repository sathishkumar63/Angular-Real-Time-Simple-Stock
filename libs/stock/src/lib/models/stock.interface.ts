export interface IStock {
  name?: string;
  symbol: string;
  currency?: string;
  currentPrice: number;
  dayHighPrice?: number;
  dayLowPrice?: number;
  fiftyTwoWeekLow?: number;
  fiftyTwoWeekHigh?: number;
  regularMarketVolume?: number;
  isActive?: boolean;
}
