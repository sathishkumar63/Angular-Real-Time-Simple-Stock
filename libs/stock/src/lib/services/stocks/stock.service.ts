import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';

@Injectable({
  providedIn: 'root',
})
export class StockService {
  constructor(private http: HttpClient) {}

  getStocks(symbols: string[]): Observable<any> {
    // HttpRequest has been blocked by CORS policy. So, add cors-anywhere.herokuapp.com to this link.
    const uri = `https://cors-anywhere.herokuapp.com/query1.finance.yahoo.com/v7/finance/quote?symbols=${symbols?.join()}`;
    return this.http.get(uri);
  }

  formatStocksResponse(stocks: any[]): any[] {
    return stocks?.map((stock) => ({
      name: stock.displayName,
      symbol: stock.symbol,
      currency: stock.currency,
      currentPrice: stock.regularMarketPrice,
      dayHighPrice: stock.regularMarketDayHigh,
      dayLowPrice: stock.regularMarketDayLow,
      fiftyTwoWeekLow: stock.fiftyTwoWeekLow,
      fiftyTwoWeekHigh: stock.fiftyTwoWeekHigh,
      regularMarketVolume: stock.regularMarketVolume
    }));
  }
}
