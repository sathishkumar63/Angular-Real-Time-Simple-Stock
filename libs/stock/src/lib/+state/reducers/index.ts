import { ActionReducerMap } from '@ngrx/store';
import * as fromStocks from './stocks.reducer';

export * from './stocks.reducer';
export const stocksFeatureKey = 'simpleStocks';

export interface StocksFeatureState {
  stocks: fromStocks.StocksState;
}

export const reducers: ActionReducerMap<StocksFeatureState> = {
  stocks: fromStocks.stocksReducer,
};
