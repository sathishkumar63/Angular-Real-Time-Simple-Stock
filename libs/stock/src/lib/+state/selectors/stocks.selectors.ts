import { createFeatureSelector, createSelector } from '@ngrx/store';
import { IStock } from '../../models';
import * as fromStocks from '../reducers';

export const selectStocksState = createFeatureSelector<
  fromStocks.StocksFeatureState
>(fromStocks.stocksFeatureKey);

export const getStocksState = createSelector(
  selectStocksState,
  (state: fromStocks.StocksFeatureState) => state?.stocks
);

export const getStocksData = createSelector(
  getStocksState,
  (state: fromStocks.StocksState): IStock[] => state?.data
);

export const isStocksLoaded = createSelector(
  getStocksState,
  (state: fromStocks.StocksState): boolean => state?.loaded
);

export const isStocksLoading = createSelector(
  getStocksState,
  (state: fromStocks.StocksState): boolean => state?.loading
);
