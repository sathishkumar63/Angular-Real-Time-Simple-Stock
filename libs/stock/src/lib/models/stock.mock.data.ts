import { StocksFeatureState, StocksState } from '../+state';
import { IStock } from './stock.interface';

export const mockStocks: IStock[] = [
  {
    name: 'Alphabet',
    symbol: 'GOOG',
    currency: 'USD',
    currentPrice: 2228.55,
    dayHighPrice: 2269.94,
    dayLowPrice: 2217.22,
    fiftyTwoWeekLow: 2044.16,
    fiftyTwoWeekHigh: 3042,
    regularMarketVolume: 1567487,
  },
  {
    name: 'Microsoft',
    symbol: 'MSFT',
    currency: 'USD',
    currentPrice: 252.99,
    dayHighPrice: 260.54,
    dayLowPrice: 252.86,
    fiftyTwoWeekLow: 246.44,
    fiftyTwoWeekHigh: 349.67,
    regularMarketVolume: 31445841,
  },
];

export const mockStocksState: StocksState = {
  data: mockStocks,
  loading: false,
  loaded: true
};

export const mockStocksFeatureState: StocksFeatureState = {
  stocks: mockStocksState
};
