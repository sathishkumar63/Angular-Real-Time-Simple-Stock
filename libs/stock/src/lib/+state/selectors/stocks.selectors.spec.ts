import { mockStocks, mockStocksFeatureState } from '../../models';
import { initialState, StocksState } from '../reducers';
import * as stocksSelectors from './stocks.selectors';

describe('Stocks Selectors', () => {
  it('should select stocks feature state', () => {
    // arrange + act
    const result = stocksSelectors.selectStocksState.projector(
      mockStocksFeatureState
    );
    // assert
    expect(result).toMatchSnapshot();
  });

  it('should return stocks data', () => {
    // arrange
    const mockData: StocksState = {
      ...initialState,
      data: mockStocks,
    };
    // act
    const result = stocksSelectors.getStocksData.projector(mockData);
    // assert
    expect(result).toEqual(mockStocks);
    expect(result).toMatchSnapshot();
  });

  it('should return loaded status', () => {
    // arrange
    const mockData: StocksState = {
      ...initialState,
      data: mockStocks,
    };
    // act
    const result = stocksSelectors.isStocksLoaded.projector(mockData);
    // assert
    expect(result).toMatchSnapshot();
  });

  it('should return loading status', () => {
    // arrange
    const mockData: StocksState = {
      ...initialState,
      data: mockStocks,
    };
    // act
    const result = stocksSelectors.isStocksLoading.projector(mockData);
    // assert
    expect(result).toMatchSnapshot();
  });
});
