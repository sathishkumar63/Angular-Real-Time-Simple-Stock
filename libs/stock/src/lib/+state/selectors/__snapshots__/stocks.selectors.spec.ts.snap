// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stocks Selectors should return loaded status 1`] = `false`;

exports[`Stocks Selectors should return loading status 1`] = `false`;

exports[`Stocks Selectors should return stocks data 1`] = `
Array [
  Object {
    "currency": "USD",
    "currentPrice": 2228.55,
    "dayHighPrice": 2269.94,
    "dayLowPrice": 2217.22,
    "fiftyTwoWeekHigh": 3042,
    "fiftyTwoWeekLow": 2044.16,
    "name": "Alphabet",
    "regularMarketVolume": 1567487,
    "symbol": "GOOG",
  },
  Object {
    "currency": "USD",
    "currentPrice": 252.99,
    "dayHighPrice": 260.54,
    "dayLowPrice": 252.86,
    "fiftyTwoWeekHigh": 349.67,
    "fiftyTwoWeekLow": 246.44,
    "name": "Microsoft",
    "regularMarketVolume": 31445841,
    "symbol": "MSFT",
  },
]
`;

exports[`Stocks Selectors should select stocks feature state 1`] = `
Object {
  "stocks": Object {
    "data": Array [
      Object {
        "currency": "USD",
        "currentPrice": 2228.55,
        "dayHighPrice": 2269.94,
        "dayLowPrice": 2217.22,
        "fiftyTwoWeekHigh": 3042,
        "fiftyTwoWeekLow": 2044.16,
        "name": "Alphabet",
        "regularMarketVolume": 1567487,
        "symbol": "GOOG",
      },
      Object {
        "currency": "USD",
        "currentPrice": 252.99,
        "dayHighPrice": 260.54,
        "dayLowPrice": 252.86,
        "fiftyTwoWeekHigh": 349.67,
        "fiftyTwoWeekLow": 246.44,
        "name": "Microsoft",
        "regularMarketVolume": 31445841,
        "symbol": "MSFT",
      },
    ],
    "loaded": true,
    "loading": false,
  },
}
`;
